core/src/test/scala/unit/kafka/server/ServerStartupTest.scala

```diff
@@ -86,22 +86,32 @@ class ServerStartupTest extends ZooKeeperTestHarness {
   @Test
   def testBrokerStateRunningAfterZK {
     val brokerId = 0
-    val mockBrokerState = EasyMock.niceMock(classOf[kafka.server.BrokerState])
 
     class BrokerStateInterceptor() extends BrokerState {
+
       override def newState(newState: BrokerStates): Unit = {
         val brokers = zkUtils.getAllBrokersInCluster()
         assertEquals(1, brokers.size)
         assertEquals(brokerId, brokers.head.id)
+        super.newState(newState)
       }
     }
 
+    //Enforce call order
+    val mockBrokerState = EasyMock.strictMock(classOf[kafka.server.BrokerState])
+    val brokerStateInterceptor = new BrokerStateInterceptor
+
     class MockKafkaServer(override val config: KafkaConfig, override val brokerState: BrokerState = mockBrokerState) extends KafkaServer(config) {}
 
     val props = TestUtils.createBrokerConfig(brokerId, zkConnect)
     val server = new MockKafkaServer(KafkaConfig.fromProps(props))
 
-    EasyMock.expect(mockBrokerState.newState(RunningAsBroker)).andDelegateTo(new BrokerStateInterceptor).once()
+    EasyMock.expect(mockBrokerState.newState(Starting)).once()
+    EasyMock.expect(mockBrokerState.newState(RecoveringFromUncleanShutdown)).once()
+    EasyMock.expect(mockBrokerState.newState(RunningAsBroker)).andDelegateTo(brokerStateInterceptor).once()
+    EasyMock.expect(mockBrokerState.newState(RunningAsController)).andDelegateTo(brokerStateInterceptor).once()
+    EasyMock.expect(mockBrokerState.newState(BrokerShuttingDown)).once()
+    EasyMock.expect(mockBrokerState.newState(NotRunning)).once()
```

Comment:
It may be a good idea to have a second broker that is not a controller and verify the behaviour for that too.